go-controller/pkg/ovn/controller/services/repair.go

```diff
@@ -2,154 +2,175 @@ package services
 
 import (
 	"fmt"
+	"sync"
+	"sync/atomic"
 	"time"
 
-	"github.com/pkg/errors"
-
+	globalconfig "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/acl"
-	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/gateway"
-	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/loadbalancer"
+	ovnlb "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/loadbalancer"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
 
-	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/util/sets"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
 	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
 	"k8s.io/klog/v2"
 )
 
-// Repair is a controller loop that can work as in one-shot or periodic mode
-// it checks the OVN Load Balancers and delete the ones that doesn't exist in
+// repair handles pre-sync and post-sync service cleanup.
+// It checks the OVN Load Balancers and delete the ones that doesn't exist in
 // the Kubernetes Service Informer cache.
+// It also deletes the legacy load balancers once a full sync run has completed.
 // Based on:
 // https://raw.githubusercontent.com/kubernetes/kubernetes/release-1.19/pkg/registry/core/service/ipallocator/controller/repair.go
-type Repair struct {
-	interval time.Duration
-	// serviceTracker tracks services and maps them to OVN LoadBalancers
-	serviceLister        corelisters.ServiceLister
-	clusterPortGroupUUID string
+type repair struct {
+	sync.Mutex
+	serviceLister corelisters.ServiceLister
+
+	// We want to run some functions after every service is successfully synced, so tpoplate this
+	// list with every service that should be in the informer queue before we start the ServiceController
+	// workers.
+	unsyncedServices sets.String
+
+	// Really a boolean, but an int32 for atomicitiy purposes
+	semLegacyLBsDeleted uint32
 }
 
 // NewRepair creates a controller that periodically ensures that there is no stale data in OVN
-func NewRepair(interval time.Duration, serviceLister corelisters.ServiceLister, clusterPortGroupUUID string) *Repair {
-	return &Repair{
-		interval:             interval,
-		serviceLister:        serviceLister,
-		clusterPortGroupUUID: clusterPortGroupUUID,
+func newRepair(serviceLister corelisters.ServiceLister) *repair {
```

Comment:
We cannot delete the legacy LBs until we've successfully synced every service at least once. Otherwise, we will disrupt all new connections. It's not a question of blocking, it's a question of availability.

Think of it this way: if vip A is on the legacy LBs, we need to migrate it before we can delete them. Once we've synced every service at least once, then it's safe to delete the legacy LBs.

Reply:
I can agree with you that it would be more efficient to remove the legacy LB as you migrate to the new one. That is a pattern we dont typically do in ovn-k, but it's something that is good to move to. However, from what I see in the patch it looks awkward, and I'm not sure it is doing as intended. For example, in runBeforeSync:

```
	// Look for any load balancers whose Service no longer exists in the apiserver
	staleLBs := []string{}
	for _, lb := range existingLBs {
		// Extract namespace + name, look to see if it exists
		owner := lb.ExternalIDs["k8s.ovn.org/owner"]
		namespace, name, err := cache.SplitMetaNamespaceKey(owner)
		if err != nil || namespace == "" {
			klog.Warningf("Service LB %#v has unreadable owner, deleting", lb)
			staleLBs = append(staleLBs, lb.UUID)
		}

		_, err = r.serviceLister.Services(namespace).Get(name)
		if apierrors.IsNotFound(err) {
			klog.V(5).Infof("Found stale service LB %#v", lb)
			staleLBs = append(staleLBs, lb.UUID)
		}
	}

	// Delete those stale load balancers
	if err := ovnlb.DeleteLBs(nil, staleLBs); err != nil {
		klog.Errorf("Failed to delete stale LBs: %v", err)
		return fmt.Errorf("failed to delete stale LBs: %w", err)
	}
```

wont you range on existingLBs, get an error for a legacy lb because you cant SplitMetaNamespaceKey, and then delete it here?

2nd thing, is I'm not sure why you are deleting all the legacy ones in a goroutine. There are several problems with this:
a. you dont remove all the legacy ones until all of the services sync - you could have conflicting LBs configured for some time - would be better to identify the conflict (via vip/port) and delete that when you sync the same service
b. if you cant remove the conflicting legacy lb, this should be a fail to add the service that conflicted with this VIP, and the add/update event should be re-queued 
c. in runBeforeSync you should identify any LBs that do not correspond to a current svc VIP/port, and those are all legacy and safe to remove

Reply:
> wont you range on existingLBs, get an error for a legacy lb because you cant SplitMetaNamespaceKey, and then delete it here?

No, because I search for all Load_Balancers with an ExternalID of `"k8s.ovn.org/kind": "Service"`. The legacy Load_Balancer rows won't have this.

> 2nd thing, is I'm not sure why you are deleting all the legacy ones in a goroutine. There are several problems with this:
> a. you dont remove all the legacy ones until all of the services sync - you could have conflicting LBs configured for some time - would be better to identify the conflict (via vip/port) and delete that when you sync the same service

That would be a potential problem, but the code removes individual vips from the legacy load balancers as the Services are synced, so we're fine there. This part is only concerned with deleting the legacy LBs after migration.

> b. if you cant remove the conflicting legacy lb, this should be a fail to add the service that conflicted with this VIP, and the add/update event should be re-queued

What do you mean by conflicting?

> c. in runBeforeSync you should identify any LBs that do not correspond to a current svc VIP/port, and those are all legacy and safe to remove

I already do that for non-legacy LBs, and the legacy LBs are just deleted altogether. What would that get us?